File: old-server/src/main/java/org/camposmdev/server/model/ClientRegistry.java
package org.camposmdev.server.model;

import io.vertx.core.json.JsonObject;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

public class ClientRegistry {
    private static ClientRegistry cr;

    public static ClientRegistry get() {
        if (cr == null)
            cr = new ClientRegistry();
        return cr;
    }

    private final List<WSAgent> clients;

    private ClientRegistry() {
        this.clients = Collections.synchronizedList(new LinkedList<>());
    }

    public void add(WSAgent x) {
        System.out.println("Registered client " + x);
        clients.add(x);
    }

    public void remove(WSAgent x) {
        System.out.println("Removed client " + x);
        clients.remove(x);
    }

    public int size() {
        return clients.size();
    }

//    public boolean isPlayerTaken(String name) {
//        return clients.stream()
//                .filter(x -> x.getPlayer() != null && x.getPlayer().getName().equals(name))
//                .toList().size() >= 1;
//    }

    public WSAgent getById(String id) {
        return clients.stream().filter(x -> x.getUserId().equals(id)).findFirst().orElse(null);
    }

    public void sendMessageTo(String id, JsonObject arg) {
        getById(id).getWS().writeTextMessage(arg.toString());
    }

    /**
     * Sends a message to global chat
     * @param arg Message to be sent to all online users
     */
    public void notifyAll(JsonObject arg) {
        for (WSAgent c : clients) {
            c.getWS().writeTextMessage(arg.toString());
        }
    }

    @Override
    public String toString() {
        return this.clients.toString();
    }
}

File: fs-core/fs-util/src/main/java/io/github/camposmdev/foursouls/core/util/ClientRegistry.kt
package io.github.camposmdev.foursouls.core.util

import java.util.*
import java.util.function.Predicate

abstract class ClientRegistry<T: AbstractServerWSClient<U>, U> {
    val clients: MutableList<T> by lazy {
        Collections.synchronizedList(LinkedList())
    }

    open fun add(client: T): Boolean {
        return clients.add(client)
    }

    fun remove(filter: Predicate<T>): Boolean {
        return clients.removeIf(filter)
    }

    fun size(): Int {
        return clients.size
    }

    fun isEmpty(): Boolean {
        return clients.size <= 0
    }
}